### ISSUE TITLE
Problems with page “Resize CPU and Memory Resources assigned to Containers”

### ISSUE THREAD

**Issue by pinakig22** [kubernetes/website#45079]
While testing the task mentioned in [](https://kubernetes.io/docs/tasks/configure-pod-container/resize-container-resources/), the actual behaviour is not as per the doc (differences outlined below). I used [killercoda](https://killercoda.com/playgrounds/scenario/kubernetes) to perform the tasks.

https://github.com/kubernetes/website/issues/41365#issuecomment-1566976281 mentions the prerequisite section should have mention of `InPlacePodVerticalScaling` enabled on the control plane and on nodes, but it is not in document [v1.27](https://v1-27.docs.kubernetes.io/docs/tasks/configure-pod-container/resize-container-resources/) and above.

Kindly help to update the document so the task can be successfully completed.

K8 version is 1.29
```
controlplane $ kubectl get nodes    
NAME           STATUS   ROLES           AGE     VERSION
controlplane   Ready    control-plane   5d19h   v1.29.0
node01         Ready    <none>          5d19h   v1.29.0
controlplane $ 
```

1. The pod created using `kubectl create -f https://k8s.io/examples/pods/qos/qos-pod-5.yaml` doesn't result in addition of `resizePolicy[*].restartPolicy` defaulted to `NotRequired` to `spec.containers`when checked using `kubectl get pod qos-demo-5 --output=yaml --namespace=qos-example`, as per example output in the doc. 

![image](https://github.com/kubernetes/website/assets/36029504/75e8cf1e-c80a-4dbe-9647-6f1080269fe7)

```
controlplane $ kubectl get pod qos-demo-5 --output=yaml --namespace=qos-example|grep "resizePolicy"
controlplane $
```
2. There is no mention of `allocatedResources` in the output yaml as per the doc

![image](https://github.com/kubernetes/website/assets/36029504/4262be1c-b6b6-446e-a767-63d06e429275)

```
controlplane $ kubectl get pod qos-demo-5 --output=yaml --namespace=qos-example|grep "allocatedResources"
controlplane $ 
```
### Output of steps as below
```
controlplane $ kubectl create namespace qos-example
namespace/qos-example created
controlplane $ kubectl create -f https://k8s.io/examples/pods/qos/qos-pod-5.yaml
pod/qos-demo-5 created
controlplane $ kubectl get pod qos-demo-5 --output=yaml --namespace=qos-example
apiVersion: v1
kind: Pod
metadata:
  annotations:
    cni.projectcalico.org/containerID: 21a131cf0e926780bd6d1b2ae4dc82ccfff8a6cad68a6f62207ff07bbf6954d0
    cni.projectcalico.org/podIP: 192.168.1.4/32
    cni.projectcalico.org/podIPs: 192.168.1.4/32
  creationTimestamp: "2024-02-10T02:57:38Z"
  name: qos-demo-5
  namespace: qos-example
  resourceVersion: "5008"
  uid: ee4278b5-bf3a-402a-837b-bdb04e9fa4d1
spec:
  containers:
  - image: nginx
    imagePullPolicy: Always
    name: qos-demo-ctr-5
    resources:
      limits:
        cpu: 700m
        memory: 200Mi
      requests:
        cpu: 700m
        memory: 200Mi
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-hjww8
      readOnly: true
  dnsPolicy: ClusterFirst
  enableServiceLinks: true
  nodeName: node01
  preemptionPolicy: PreemptLowerPriority
  priority: 0
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext: {}
  serviceAccount: default
  serviceAccountName: default
  terminationGracePeriodSeconds: 30
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  volumes:
  - name: kube-api-access-hjww8
    projected:
      defaultMode: 420
      sources:
      - serviceAccountToken:
          expirationSeconds: 3607
          path: token
      - configMap:
          items:
          - key: ca.crt
            path: ca.crt
          name: kube-root-ca.crt
      - downwardAPI:
          items:
          - fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
            path: namespace
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2024-02-10T02:57:38Z"
    status: "False"
    type: PodReadyToStartContainers
  - lastProbeTime: null
    lastTransitionTime: "2024-02-10T02:57:38Z"
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: "2024-02-10T02:57:38Z"
    message: 'containers with unready status: [qos-demo-ctr-5]'
    reason: ContainersNotReady
    status: "False"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: "2024-02-10T02:57:38Z"
    message: 'containers with unready status: [qos-demo-ctr-5]'
    reason: ContainersNotReady
    status: "False"
    type: ContainersReady
  - lastProbeTime: null
    lastTransitionTime: "2024-02-10T02:57:38Z"
    status: "True"
    type: PodScheduled
  containerStatuses:
  - image: nginx
    imageID: ""
    lastState: {}
    name: qos-demo-ctr-5
    ready: false
    restartCount: 0
    started: false
    state:
      waiting:
        reason: ContainerCreating
  hostIP: 172.30.2.2
  hostIPs:
  - ip: 172.30.2.2
  phase: Pending
  qosClass: Guaranteed
  startTime: "2024-02-10T02:57:38Z"
controlplane $ 
```

3. While trying to resize CPU request & limit using the command `kubectl -n qos-example patch pod qos-demo-5 --patch '{"spec":{"containers":[{"name":"qos-demo-ctr-5", "resources":{"requests":{"cpu":"800m"}, "limits":{"cpu":"800m"}}}]}}'` results in below error which suggests the cpu limit/request values can't be changed, and lists out what can be changed.

```
controlplane $ kubectl -n qos-example patch pod qos-demo-5 --patch '{"spec":{"containers":[{"name":"qos-demo-ctr-5", "resources":{"requests":{"cpu":"800m"}, "limits":{"cpu":"800m"}}}]}}'
The Pod "qos-demo-5" is invalid: spec: Forbidden: pod updates may not change fields other than `spec.containers[*].image`,`spec.initContainers[*].image`,`spec.activeDeadlineSeconds`,`spec.tolerations` (only additions to existing tolerations),`spec.terminationGracePeriodSeconds` (allow it to be set to 1 if it was previously negative)
  core.PodSpec{
        Volumes:        {{Name: "kube-api-access-hjww8", VolumeSource: {Projected: &{Sources: {{ServiceAccountToken: &{ExpirationSeconds: 3607, Path: "token"}}, {ConfigMap: &{LocalObjectReference: {Name: "kube-root-ca.crt"}, Items: {{Key: "ca.crt", Path: "ca.crt"}}}}, {DownwardAPI: &{Items: {{Path: "namespace", FieldRef: &{APIVersion: "v1", FieldPath: "metadata.namespace"}}}}}}, DefaultMode: &420}}}},
        InitContainers: nil,
        Containers: []core.Container{
                {
                        ... // 6 identical fields
                        EnvFrom: nil,
                        Env:     nil,
                        Resources: core.ResourceRequirements{
                                Limits: core.ResourceList{
-                                       s"cpu":    {i: resource.int64Amount{value: 700, scale: -3}, s: "700m", Format: "DecimalSI"},
+                                       s"cpu":    {i: resource.int64Amount{value: 800, scale: -3}, s: "800m", Format: "DecimalSI"},
                                        s"memory": {i: {...}, Format: "BinarySI"},
                                },
                                Requests: core.ResourceList{
-                                       s"cpu":    {i: resource.int64Amount{value: 700, scale: -3}, s: "700m", Format: "DecimalSI"},
+                                       s"cpu":    {i: resource.int64Amount{value: 800, scale: -3}, s: "800m", Format: "DecimalSI"},
                                        s"memory": {i: {...}, Format: "BinarySI"},
                                },
                                Claims: nil,
                        },
                        ResizePolicy:  nil,
                        RestartPolicy: nil,
                        ... // 13 identical fields
                },
        },
        EphemeralContainers: nil,
        RestartPolicy:       "Always",
        ... // 28 identical fields
  }

controlplane $
```
4. However using the below pod definition does have details for `resizePolicy`, but under **`annotations`** and **not** under `spec.containers`. But there is still **no mention of** `allocatedResources` in the yaml output. 

```
## resizePolicy ##
controlplane $ kubectl get pod qos-demo-5 --output=yaml --namespace=qos-example |grep resize
      {"apiVersion":"v1","kind":"Pod","metadata":{"annotations":{},"name":"qos-demo-5","namespace":"qos-example"},"spec":{"containers":[{"image":"nginx","name":"qos-demo-ctr-5","resizePolicy":[{"resourceName":"cpu","restartPolicy":"NotRequired"},{"resourceName":"memory","restartPolicy":"RestartContainer"}],"resources":{"limits":{"cpu":"700m","memory":"200Mi"},"requests":{"cpu":"700m","memory":"200Mi"}}}]}}
controlplane $

### allocatedResources ###
controlplane $ kubectl get pod qos-demo-5 --output=yaml --namespace=qos-example |grep allocatedResources
controlplane $ kubectl get pod qos-demo-5 --output=yaml --namespace=qos-example |grep -i allocatedResources
controlplane $
```
```
## Pod Definition ## 
apiVersion: v1
kind: Pod
metadata:
  name: qos-demo-5
  namespace: qos-example
spec:
  containers:
  - name: qos-demo-ctr-5
    image: nginx
    resizePolicy:
    - resourceName: cpu
      restartPolicy: NotRequired
    - resourceName: memory
      restartPolicy: RestartContainer
    resources:
      limits:
        memory: "200Mi"
        cpu: "700m"
      requests:
        memory: "200Mi"
        cpu: "700m"
```

5. Further the result of patching is still as above, i.e. not allowed to update the values

**Comment by sftim** [comment]
/sig node

/retitle Problems with page “Resize CPU and Memory Resources assigned to Containers”

/priority backlog
(feature is alpha)
/triage accepted

**Comment by vaibhav2107** [comment]
Is this issue fixed by PR #45350 ? If yes, then we can close this issue

**Comment by vaibhav2107** [comment]
/close
Please feel free to re-open this if something more is left
